fix: make latest query results specific to db_role

ajkerrigan opened the pull request:
[ENG-5525](https://stacklet.atlassian.net/browse/ENG-5525)

### what

Make sure that the latest query results are specific to the current user's `db_role`.

### why

Aegon encountered an issue where users with limited access were running queries, and then admins were seeing those limited cached results instead of the full result set.

### testing

This change doesn't affect which query results a user can see, or which cached results come back for a limited visibility user. The _only_ impact should be when a non-limited (admin) user tries to fetch the latest cached results for a query - they'll be guaranteed to see only non-limited (admin) cached results.

To validate that I...

- Set up 3 users with 3 different db_roles in my sandbox: limited, limited2, and admin (`None`)
- Refreshed a test query 3 times, with these results:
  - Limited user --> query result 52012
  - Admin user --> query result 52013
  - Limited2 user --> query result 52017
- Issued a series of GET requests for the test query, with these results (gist: only the admin response changes):

```
limited response from local: 52012
limited response from ajsbx: 52012
limited2 response from local: 52017
limited2 response from ajsbx: 52017
admin response from local: 52013
admin response from ajsbx: 52017
```
- Tested fetching each query result using each test user, with these results (gist: this PR changes nothing):

```
limited role fetching limited results from local: HTTP 200 response, row count: 0
limited role fetching limited results from ajsbx: HTTP 200 response, row count: 0
limited2 role fetching limited results from local: HTTP 404 response, row count: N/A
limited2 role fetching limited results from ajsbx: HTTP 404 response, row count: N/A
admin role fetching limited results from local: HTTP 200 response, row count: 0
admin role fetching limited results from ajsbx: HTTP 200 response, row count: 0
limited role fetching limited2 results from local: HTTP 404 response, row count: N/A
limited role fetching limited2 results from ajsbx: HTTP 404 response, row count: N/A
limited2 role fetching limited2 results from local: HTTP 200 response, row count: 0
limited2 role fetching limited2 results from ajsbx: HTTP 200 response, row count: 0
admin role fetching limited2 results from local: HTTP 200 response, row count: 0
admin role fetching limited2 results from ajsbx: HTTP 200 response, row count: 0
limited role fetching admin results from local: HTTP 404 response, row count: N/A
limited role fetching admin results from ajsbx: HTTP 404 response, row count: N/A
limited2 role fetching admin results from local: HTTP 404 response, row count: N/A
limited2 role fetching admin results from ajsbx: HTTP 404 response, row count: N/A
admin role fetching admin results from local: HTTP 200 response, row count: 25756
admin role fetching admin results from ajsbx: HTTP 200 response, row count: 25756
```

### docs

n/a

[ENG-5525]: https://stacklet.atlassian.net/browse/ENG-5525?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ajkerrigan committed:
chore: address some local dev papercuts

- Define `ASSETDB_DATABASE_URI` in docker-compose to avoid startup
  errors trying to determine IAM auth from a non-existent AssetDB URI

- [Hack] Duplicate the Poetry name & version keys to the top-level
  project in pyproject.toml, avoids some error cases installing
  the root package into a virtual environment

- Set the active schema before running SQL commands in the
  `make create_database` step, so we don't get "relation not found"
  errors referencing `query_results` rather than `redash.query_results`

ajkerrigan committed:
fix: specify db_role when fetching latest query results

- Make sure that `QueryResult.get_latest()` considers `db_role` when
  determining the latest available query result.

- Add a test to be sure that change is working as expected.

- Go up a level to be sure that `run_query()` also has `db_role` context
  and passes it along to `get_latest()`.

johnsca commented on redash/cli/database.py:
```diff
@@ -79,6 +79,7 @@ def create_tables():
         sqlalchemy.orm.configure_mappers()
         db.create_all()
 
+        db.session.execute(f"SET search_path to {settings.SQLALCHEMY_DATABASE_SCHEMA}")
```
This seems like a drive-by unrelated to this issue, and I don't think it's correct. This would replace the existing `search_path` which would likely break the use of some extensions we install. See [this Slack thread](https://stacklet.slack.com/archives/C0465C3E7BN/p1752696450286199?thread_ts=1752690423.739259&cid=C0465C3E7BN).

ajkerrigan replied:
Oh sure, I'm happy to take that out. I only put it in because without it I was getting a bunch of errors about not being able to find the `query_results` relation running [these operations](https://github.com/stacklet/redash/blob/f4d487d9a3ac358543ec235471d49018219d604d/redash/cli/database.py#L82-L97). Adding this line _or_ fully qualifying the table names seemed to fix the issue - I could see either fully qualifying the table names or doing nothing (since like you said it's an unrelated drive-by thing).

ajkerrigan replied:
I removed this change entirely rather than fully qualifying the `query_results` references, because I realize running against a local empty database just dies right after this anyway. I think we'd need more changes to do things like add our expected roles (`limited_visibility`, etc) and that's definitely further than I want to take this PR :laughing: .

johnsca reviewed (commented):
I think these changes seem good, but we might also need the changes I suggested in the Slack thread, specifically:
* Changing [this block](https://github.com/stacklet/redash/blob/a2e30509db5d8551893ce08c8c9cad3c4238794c/redash/models/__init__.py#L412-L420) to handle the `db_role is None` case with a `db_role.is_(None)` filter rather than just skipping it.
* Removing [this `if` check](https://github.com/stacklet/redash/blob/a2e30509db5d8551893ce08c8c9cad3c4238794c/redash/serializers/__init__.py#L140) and making the block unconditional.

I haven't dug deeply into it, though, but my thoughts are that not including those changes might make it re-run queries when it doesn't need to.

johnsca commented on redash/models/__init__.py:
```diff
@@ -363,6 +363,7 @@ def get_latest(cls, data_source, query, max_age=0, is_hash=False):
             query = cls.query.filter(
                 cls.query_hash == query_hash,
                 cls.data_source == data_source,
+                cls.db_role == db_role,
```
Oh, I just realized that this will not do what you expect when `db_role` is `None` (i.e., user is an admin). Instead of finding the results created by the admin, it will just never find any results. (`col == null` is neither `false` nor `true`). You'll need to check for a `None` value and use `cls.db_role.is_(None)`. I think this would work:

```suggestion
                cls.db_role == db_role if db_role is not None else cls.db_role.is_(None),
```

ajkerrigan replied:
This was a good call. I ended up still hitting some "this isn't quite a boolean" errors so rejiggered the statement a bit, but the core point is well taken :+1:

ajkerrigan committed:
tweak serialize_query() and QueryResult.get_latest()

Address review feedback and fine-tune based on local testing:

- Update the `latest_query_data_id` when serializing a query response,
  regardless of the active user's `db_role`.

- When finding the latest relevant query results, use a separate null
  check for `db_role`. Use the same role filtering in both branches of
  the `max_age` check.

ajkerrigan committed:
don't bother setting the schema in database.py

using an empty db will die later anyway because it assumes rls-related
roles exist and they won't on a fresh redash db.

ajkerrigan commented:
> I think these changes seem good, but we might also need the changes I suggested in the Slack thread, specifically:
> 
>     * Changing [this block](https://github.com/stacklet/redash/blob/a2e30509db5d8551893ce08c8c9cad3c4238794c/redash/models/__init__.py#L412-L420) to handle the `db_role is None` case with a `db_role.is_(None)` filter rather than just skipping it.
> 
>     * Removing [this `if` check](https://github.com/stacklet/redash/blob/a2e30509db5d8551893ce08c8c9cad3c4238794c/redash/serializers/__init__.py#L140) and making the block unconditional.
> 
> 
> I haven't dug deeply into it, though, but my thoughts are that not including those changes might make it re-run queries when it doesn't need to.

I did a bunch of side-by-side testing with my sandbox Redash compared to a local one based on this PR (putting that into the testing section of the PR description). For your first point, I think leaving that as-is is OK - my thinking is that an admin should still be able to see results from a limited role... they just shouldn't show up as the latest available. For your second point, good call :+1: . I removed the `if` check you mentioned and made sure to pass the `db_role` into the `get_latest()` call there.

ajkerrigan committed:
address cursor findings

nice robot, good robot

ajkerrigan committed:
base query stats on the per-role latest results

ajkerrigan commented on redash/serializers/__init__.py:
```diff
@@ -137,17 +137,18 @@ def serialize_query(
+        is_hash=True,
+        db_role=getattr(current_user, "db_role", None),
+    )
+    d["latest_query_data_id"] = latest_result and latest_result.id or None
```
Updating `retrieved_at` and `runtime` is a fair point, done.

The N+1 issue... I'm not sure how to avoid that. A couple things come to mind though...

1. We muck with the data model so that the `redash.queries` table has a row per query/db-role-specific-cached-result pair.
2. We add an option (similar to `with_stats`) to `serialize_query()` that is true when getting an individual query (`/api/queries/<id>` but false when listing queries in bulk (`/api/queries`). And we only replace the latest results & stats on the individual query fetch.

Option 1 feels gnarly. Option 2 feels shady, but maybe pragmatic? It would show the latest time _someone_ refreshed the query in a list, but retrieving an individual query would still pull the "right" results :thinking: . Options 3+ haven't occurred to me.

ajkerrigan committed:
tradeoff: avoid N+1 during bulk query fetching

- Don't override latest query result data with db_role-specific results
  when bulk-fetching queries (based on `with_stats`)
- _Do_ override latest results when fetching an individual query

The goal here is to balance API impact and result accuracy/relevance.

ajkerrigan commented:
I'm not entirely sure that my approach to managing N+1 impact is sensible. But my thought process is that if we're pulling a list of queries, seeing last run stats from another db role is OK and avoids an N+1 issue. If we're looking at an individual query, we need the "best" latest result because the next step is probably fetching those cached results.
